### PR TITLE
MAYA-111898 - MaterialX UV0 defaultgeomprop to st primvar

### PR DESCRIPTION
**Pull request by JGamache-autodesk**
MaterialX has a lot of node definitions that use a UV0 defaultgeomprop to automatically connect to UV stream at index zero. Since USD is name based, and we specified a canonical naming scheme for UV sets in PR #1528, we can insert a geompropvalue primvar reader that will allow that shader to work.
MaterialX can also specify a UV stream by index using a "texcoord" geometry node. For these, we substitute a geompropvalue primvar reader that refers to the primvar name we assigned for that index: UV2 -> st2

### CODE REVIEW

File: lib/mayaUsd/render/vp2RenderDelegate/material.cpp [hunk 1]
```diff
@@ -164,9 +164,108 @@ TF_DEFINE_PRIVATE_TOKENS(
...
+    (ND_position_vector3)
+    (ND_normal_vector3)
+    (ND_tangent_vector3)
+    (ND_bitangent_vector3)
+    // Topo affecting nodes due to channel index. We remap to geomprop in _AddMissingTexcoordReaders
+    (ND_texcoord_vector2)
+    (ND_texcoord_vector3)
+    // Color at vertices also affect topo, but we have not locked a naming scheme to go from index
+    // based to name based as we did for UV sets. We will mark them as topo-affecting, but there is
+    // nothing we can do to link them correctly to a primvar without specifying a naming scheme.
```
Comment (JGamache-autodesk) [inline lib/mayaUsd/render/vp2RenderDelegate/material.cpp:189]: Should we also specify a primvar naming scheme for these as well?

File: test/testSamples/MaterialX/MtlxUVStreamTest.usda [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+					float inputs:in1.connect = </pCube1/Looks/standardSurface2SG/MayaNG_standardSurface2SG/noise1.outputs:out>
+					float inputs:in2 = 0.5
+					float outputs:out
+				}
+				
+				def Shader "noise1"
+				{
+                    uniform token info:id = "ND_noise2d_float"
+                    float inputs:amplitude = 0.5
+					float2 inputs:texcoord.connect = </pCube1/Looks/standardSurface2SG/MayaNG_standardSurface2SG/UV1.outputs:out>
```
Comment (JGamache-autodesk) [inline test/testSamples/MaterialX/MtlxUVStreamTest.usda:67]: Explicitly connected to a texcoord reader for UV at index 1.

File: test/testSamples/MaterialX/MtlxUVStreamTest.usda [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                    float outputs:out
+				}
+				
+				def Shader "UV1"
+				{
+                    uniform token info:id = "ND_texcoord_vector2"
+					int inputs:index = 1
```
Comment (JGamache-autodesk) [inline test/testSamples/MaterialX/MtlxUVStreamTest.usda:74]: Will be replaced with geompropvalue for "st1".

File: test/testSamples/MaterialX/MtlxUVStreamTest.usda [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                    float2 outputs:out
+				}
+				
+				def Shader "ramp1"
+				{
+                    uniform token info:id = "ND_ramptb_color3"
+                    color3f inputs:valuet = (0.0, 1.0, 0.5)
+                    color3f inputs:valueb = (1.0, 0.5, 0.0)
+                    color3f outputs:out
```
Comment (JGamache-autodesk) [inline test/testSamples/MaterialX/MtlxUVStreamTest.usda:83]: Unconnected texcoord input. Will be auto-connected to an auto-added ST_reader node that reads primvar "st".

File: test/testSamples/MaterialX/MtlxUVStreamTest.usda [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+				}
+				
+                def Shader "file1"
+                {
+                    uniform token info:id = "ND_image_color4"
+                    color4f inputs:default = (0.5, 0.5, 0.5, 1)
+                    asset inputs:file = @grid.png@
+                    string inputs:uaddressmode = "periodic"
+                    string inputs:vaddressmode = "periodic"
+                    color4f outputs:out
```
Comment (JGamache-autodesk) [inline test/testSamples/MaterialX/MtlxUVStreamTest.usda:93]: Also has an unconnected texcoord reader that will auto-connect to the same ST_reader as the ramp1 node.

File: test/testSamples/MaterialX/MtlxUVStreamTest.usda [hunk 6]
```diff
@@ -0,0 +1,107 @@
+#usda 1.0
+(
```
Comment (JGamache-autodesk) [inline test/testSamples/MaterialX/MtlxUVStreamTest.usda:2]: If the "HDVP2_DEBUG_MATERIAL" TF_DEBUG flag is set, you will see the corrected MaterialX document printed in the output:
```
<?xml version="1.0"?>
<materialx version="1.38" xmlns:xi="http://www.w3.org/2001/XInclude">
  <standard_surface name="SR_USD_Mtlx_VP2_Material" type="surfaceshader">
    <input name="base_color" type="color3" output="base_color_out" nodegraph="NG_Maya" />
  </standard_surface>
  <surfacematerial name="USD_Mtlx_VP2_Material" type="material">
    <input name="surfaceshader" type="surfaceshader" nodename="SR_USD_Mtlx_VP2_Material" />
  </surfacematerial>
  <nodegraph name="NG_Maya">
    <mix name="N1" type="color3">
      <input name="bg" type="color3" nodename="N6" />
      <input name="fg" type="color3" nodename="N5" />
      <input name="mix" type="float" nodename="N2" />
    </mix>
    <swizzle name="N6" type="color3">
      <input name="channels" type="string" value="rgb" />
      <input name="in" type="color4" nodename="N7" />
    </swizzle>
    <image name="N7" type="color4">
      <input name="texcoord" type="vector2" defaultgeomprop="UV0" uiname="Texture Coordinates" nodename="ST_reader" />
    </image>
    <ramptb name="N5" type="color3">
      <input name="texcoord" type="vector2" defaultgeomprop="UV0" nodename="ST_reader" />
    </ramptb>
    <add name="N2" type="float">
      <input name="in1" type="float" nodename="N3" />
    </add>
    <noise2d name="N3" type="float">
      <input name="texcoord" type="vector2" nodename="N4" />
    </noise2d>
    <output name="base_color_out" type="color3" nodename="N1" />
    <geompropvalue name="ST_reader" type="vector2">
      <input name="geomprop" type="string" value="st" uniform="true" />
    </geompropvalue>
    <geompropvalue name="N4" type="vector2">
      <input name="geomprop" type="string" value="st1" uniform="true" />
    </geompropvalue>
  </nodegraph>
</materialx>
```

File: lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp [hunk 7]
```diff
@@ -2073,17 +2074,43 @@ void HdVP2Mesh::_UpdateDrawItem(
...
+                        if (primvarBuffer) { // this filters out the separate color & alpha entries
+                            vertexBuffers.addBuffer(p.GetText(), primvarBuffer);
+                        }
+                        addedPrimvars.insert(p);
+                    };
+
+                // Points and normals always are at the beginning of vertex requirements:
+                addPrimvar(HdTokens->points);
+                addPrimvar(HdTokens->normals);
+                // Then add required primvars *in order*:
```
Comment (JGamache-autodesk) [inline lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp:2102]: The MaterialX unit test had st and st1 swapped 50% of the time. This mkes sure the order stays correct.

**Review by williamkrick — APPROVED** [review]

File: lib/mayaUsd/render/vp2RenderDelegate/material.cpp (hunk 1 above)
Comment (kxl-adsk) [reply lib/mayaUsd/render/vp2RenderDelegate/material.cpp:189]: Is there any scheme existing out there already?

**Review by kxl-adsk — APPROVED** [review]

**Pull request merged**